IPIP-379: Delegated IPNS HTTP API (#379)
ipfs/specs: IPIP/0379-delegated-ipns-http-api.md

## Design rationale

The rationale for delegated IPNS over HTTP APIs closely follows the reasoning listed in [IPIP-337](0337-delegated-routing-http-api.md#design-rationale).

The document proposes the following:
- Use of HTTP semantics for publication and resolution of naming records.
- Preference for human-readable request and response encoding, such as JSON format
- Optional backward compatibility support for the existing ProtocolBuffer format using `Content-Type: application/vnd.ipfs.ipns-record`
- Use of extra headers in `OPTIONS` response to communicate the supported capabilities and limitations, namely:
  - `X-Ipns-Allow-Max-Size` -- to signal maximum supported IPNS record size
  - `X-Ipns-Allow-Protobuf` -- to signal whether the server supports ProtocolBuffer formatted records.
- Streaming interaction is not supported.

### User benefit

The ability of offload naming onto another process or server via an idiomatic and simple HTTP API opens up an opportunity to not only reduce the burden of naming system management, but also introduce alternative naming systems across other routing systems, namely IPNI.

> ### Compatibility

Reviewer: TODO(before we ratify this):
- [ ] clients use `application/vnd.ipfs.ipns-record` ([IANA](https://github.com/ipfs/in-web-browsers/issues/209), [multicodec](https://github.com/multiformats/multicodec/pull/312)) no matter what router type is used for GET/PUT of IPNS record
- [ ] clients can use trustless gateway and  /routing/v1 endpoints  interchangeably, and a single service can share IPNS record cache across all supported routers DHT/Pubsub/HTTP

#### Backwards Compatibility

##### Serialization Format

> IPNS records use [ProtocolBuffer serialisation format](../ipns/IPNS.md#record-serialization-format).

Reviewer: ```suggestion
IPNS records use [`application/vnd.ipfs.ipns-record` protobuf serialization format](https://specs.ipfs.tech/ipns/ipns-record/#record-serialization-format).
```

This format is widely in use in IPNS over PubSub and DHT routing systems.
One of the motivations of this document is to introduce simple to use HTTP APIs and ultimately reduce barrier for interaction across alternative systems.
Further, interoperability across the existing and HTTP APIs is also desirable in order to reduce the barrier for adoption of the delegated HTTP APIs.

The specification here maintains backwards compatibility in terms of record serialisation, with preference for human-readable formats such as JSON.

> To maximize interoperability with existing ecosystem, the canonical [IPNS record serialization format](../ipns/IPNS.md#record-serialization-format) (`0x0300`) can be requested with content type `application/vnd.ipfs.ipns-record`.

Reviewer: ```suggestion
To maximize interoperability with existing ecosystem, the canonical [IPNS record serialization format](https://specs.ipfs.tech/ipns/ipns-record/#record-serialization-format) (`0x0300`) can be requested with content type `application/vnd.ipfs.ipns-record`.
```

> ##### Reframe
> 
> See [IPIP-337/Backwards Compatibility](0337-delegated-routing-http-api.md#backwards-compatibility).
> 
> #### Forwards Compatibility
> 
> See [IPIP-337/Forwards Compatibility](0337-delegated-routing-http-api.md#forwards-compatibility).

Reviewer: ```suggestion
#### Forwards Compatibility

See [IPIP-337/Forwards Compatibility](https://specs.ipfs.tech/ipips/ipip-0337/#forwards-compatibility).
```

### Security

All interaction over the APIs should use TLS to protect against third-party observation and tampering.

> Additionally, the IPNS records are signed by the publisher's identity and contain sequence number to avoid replay attacks.

Reviewer: ```suggestion
Additionally, the IPNS records must be [validated according to the rules stated in the IPNS spec](https://specs.ipfs.tech/ipns/ipns-record/#record-verification) before further processing.
```

> To avoid Denial of Service attack, maximum IPNS record size of `10 KiB` applies.

Reviewer: ```suggestion
To avoid Denial of Service attack, [maximum IPNS record size defined in IPNS spec](https://specs.ipfs.tech/ipns/ipns-record/#record-size-limit) applies.
```

Implements are permitted to set a lower limit. If lower than the default maximum, the limit should be discoverable via `OPTIONS` request with header key `X-Ipns-Allow-Max-Size` with value specified as the number of bytes.

> Similarly, a client may check if a server supports ProtocolBuffer formatted records by checking the `X-Ipns-Allow-Protobuf` header key in response to `OPTIONS` request. If present the header value must be either `true` or `false` the absence of the header indicates that ProtocolBuffer formatted records are not supported.

Reviewer: Since `application/vnd.ipfs.ipns-record` is mandatory, should we simplify spec and remove `Ipns-Allow-Protobuf`?

```suggestion
```

> Implements are permitted to set a lower limit. If lower than the default maximum, the limit should be discoverable via `OPTIONS` request with header key `X-Ipns-Allow-Max-Size` with value specified as the number of bytes.
> 
> Similarly, a client may check if a server supports ProtocolBuffer formatted records by checking the `X-Ipns-Allow-Protobuf` header key in response to `OPTIONS` request. If present the header value must be either `true` or `false` the absence of the header indicates that ProtocolBuffer formatted records are not supported.

Reviewer: Unnecessary complexity, as explained in the other comment.
```suggestion
Implements are permitted to set a lower limit. If lower than the default maximum, the limit should be discoverable via `OPTIONS` request with header key `X-Ipns-Allow-Max-Size` with value specified as the number of bytes.

Similarly, a client may check if a server supports ProtocolBuffer formatted records by checking the `X-Ipns-Allow-Protobuf` header key in response to `OPTIONS` request. If present the header value must be either `true` or `false` the absence of the header indicates that ProtocolBuffer formatted records are not supported.
```

Privacy in delegated IPNS is out of scope for this work.

> - The usual JSON parsing rules apply. To prevent potential Denial of Service (DoS) attack, clients should ignore responses larger than 100 providers and introduce a byte size limit that is applicable to their use case.

Reviewer: I think this is copy-paste from older IPIP  – mind rephrasing it to refer to Max Size or something?

> - The usual JSON parsing rules apply. To prevent potential Denial of Service (DoS) attack, clients should ignore responses larger than 100 providers and introduce a byte size limit that is applicable to their use case.

Reviewer: I think this is copy-paste from older IPIP  – removing.
```suggestion
- The usual JSON parsing rules apply. To prevent potential Denial of Service (DoS) attack, clients should ignore responses larger than 100 providers and introduce a byte size limit that is applicable to their use case.
```

### Alternatives

See:

> - [Libp2p-based IPNS](https://github.com/ipfs/go-ipns).

Reviewer: There is no libp2p there – records can be published and resolved over any medium, including sneakernet.

```suggestion
- [go-ipns](https://github.com/ipfs/go-ipns).
```

> - Reframe; find out more on [IPIP-337 Design Rationale](0337-delegated-routing-http-api.md#design-rationale).

Reviewer: ```suggestion
- Reframe; it was deprecated and now removed. Historical reasons can be found in [IPIP-337 Design Rationale](https://specs.ipfs.tech/ipips/ipip-0337/#design-rationale).
```
